File: app/src/main/java/fly/speedmeter/grub/MainActivity.java
package fly.speedmeter.grub;

import android.Manifest;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.GpsSatellite;
import android.location.GpsStatus;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.SystemClock;
import androidx.core.app.ActivityCompat;
import androidx.appcompat.app.AppCompatActivity;
import android.text.TextUtils;
import android.os.Bundle;
//import androidx.appcompat.widget.Toolbar;
import android.text.SpannableString;
import android.text.style.RelativeSizeSpan;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Chronometer;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.appbar.MaterialToolbar;
import com.google.gson.Gson;

public class MainActivity extends AppCompatActivity implements LocationListener,
                                  PositioningStatus.Listener {

    private SharedPreferences sharedPreferences;
    private LocationManager mLocationManager;
    private static Data data;

    private MaterialToolbar toolbar;
    private FloatingActionButton fab;
    private Menu optionsMenu;
    //private ProgressBarCircularIndeterminate progressBarCircularIndeterminate;
    private TextView satellite;
    private TextView accuracy;
    private TextView currentSpeed;
    private TextView maxSpeed;
    private TextView averageSpeed;
    private TextView distance;
    private Chronometer time;
    private Data.OnGpsServiceUpdate onGpsServiceUpdate;
    private PositioningStatus mPositioningStatus;

    private boolean firstfix;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        data = new Data(onGpsServiceUpdate);

        sharedPreferences = androidx.preference.PreferenceManager.getDefaultSharedPreferences(getApplicationContext());

        toolbar = (MaterialToolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setVisibility(View.INVISIBLE);

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[] {Manifest.permission.ACCESS_FINE_LOCATION}, 1337);
        }

        onGpsServiceUpdate = new Data.OnGpsServiceUpdate() {
            @Override
            public void update() {
                boolean autoAverage = sharedPreferences.getBoolean("auto_average", false);
                boolean imperial = sharedPreferences.getBoolean("miles_per_hour", false);
                
                double maxSpeedTemp = data.getMaxSpeed();
                double distanceTemp = data.getDistance();
                                
                double averageTemp = (autoAverage ? data.getAverageSpeedMotion() :
                                                   data.getAverageSpeed());

                String speedUnits = (imperial ? "mi/h" : "km/h");
                String distanceUnits = (imperial ? "mi" :
                                                   (distanceTemp <= 1000.0 ? "m" : "km"));
                if (imperial) {
                    maxSpeedTemp *= 0.62137119;
                    distanceTemp = distanceTemp / 1000.0 * 0.62137119;
                    averageTemp *= 0.62137119;
                } else {
                    if (distanceTemp >= 1000.0) {
                        distanceTemp /= 1000.0;
                    }
                }

                SpannableString s = new SpannableString(String.format("%.0f %s", maxSpeedTemp, speedUnits));
                s.setSpan(new RelativeSizeSpan(0.5f), s.length() - speedUnits.length() - 1, s.length(), 0);
                maxSpeed.setText(s);

                s = new SpannableString(String.format("%.0f %s", averageTemp, speedUnits));
                s.setSpan(new RelativeSizeSpan(0.5f), s.length() - speedUnits.length() - 1, s.length(), 0);
                averageSpeed.setText(s);

                s = new SpannableString(String.format("%.02f %s", distanceTemp, distanceUnits));
                s.setSpan(new RelativeSizeSpan(0.5f), s.length() - distanceUnits.length() - 1, s.length(), 0);
                distance.setText(s);
            }
        };

        mLocationManager = (LocationManager) this.getSystemService(Context.LOCATION_SERVICE);

        boolean imperial = sharedPreferences.getBoolean("miles_per_hour", false);

        String speedUnits = (imperial ? "mi/h" : "km/h");
        String lengthUnits = (imperial ? "mi" : "m");

        satellite = (TextView) findViewById(R.id.satelliteData);
        accuracy = (TextView) findViewById(R.id.accuracyData);

        maxSpeed = (TextView) findViewById(R.id.maxSpeedData);

        averageSpeed = (TextView) findViewById(R.id.averageSpeedData);

        distance = (TextView) findViewById(R.id.distance);
        distance.setText("---");

        time = (Chronometer) findViewById(R.id.time);
        time.setBase(SystemClock.elapsedRealtime());

        SpannableString s = new SpannableString(String.format("%.0f %s", 0.0f, speedUnits));
        s.setSpan(new RelativeSizeSpan(0.25f), s.length() - speedUnits.length() - 1, s.length(), 0);

        currentSpeed = (TextView) findViewById(R.id.currentSpeed);
        currentSpeed.setText(s);
        //progressBarCircularIndeterminate = (ProgressBarCircularIndeterminate) findViewById(R.id.progressBarCircularIndeterminate);
        
        createPositioningStatus();
    }

    public void onFabClick(View v) {
        if (!data.isRunning()) {
            fab.setImageDrawable(getResources().getDrawable(R.drawable.ic_action_pause));
            data.setRunning(true);
            time.setBase(SystemClock.elapsedRealtime() - data.getTime());
            time.start();
            data.setFirstTime(true);
            startService(new Intent(getBaseContext(), GpsServices.class));
            optionsMenu.findItem(R.id.action_refresh).setVisible(false);

        } else {
            fab.setImageDrawable(getResources().getDrawable(R.drawable.ic_action_play));
            data.setRunning(false);
            data.setTime(SystemClock.elapsedRealtime() - time.getBase());
            time.stop();
            toolbar.setTitle(R.string.app_name);
            stopService(new Intent(getBaseContext(), GpsServices.class));
            optionsMenu.findItem(R.id.action_refresh).setVisible(true);
        }
    }

    public void onRefreshClick() {
        resetData();
        stopService(new Intent(getBaseContext(), GpsServices.class));
    }

    @Override
    protected void onResume() {
        super.onResume();
        firstfix = true;
        if (!data.isRunning()) {
            Gson gson = new Gson();
            String json = sharedPreferences.getString("data", "");
            data = gson.fromJson(json, Data.class);
        }
        if (data == null) {
            data = new Data(onGpsServiceUpdate);
        } else {
            data.setOnGpsServiceUpdate(onGpsServiceUpdate);
        }


        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        
        if (mLocationManager == null) {
            return;
        }

        if (mLocationManager.getAllProviders().indexOf(LocationManager.GPS_PROVIDER) >= 0) {
            mLocationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 500, 0, this);
        } else {
            Log.w("MainActivity", "No GPS location provider found. GPS data display will not be available.");
        }

        /*if (!mLocationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            showGpsDisabledDialog();
        }

        mLocationManager.addGpsStatusListener(this);*/
        
        createPositioningStatus();
        
        if (!mPositioningStatus.isProviderEnabled()) {
            showGpsDisabledDialog();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mLocationManager != null) {
            mLocationManager.removeUpdates(this);
            //mLocationManager.removeGpsStatusListener(this);
        }
        mPositioningStatus.unregister();

        SharedPreferences.Editor prefsEditor = sharedPreferences.edit();
        Gson gson = new Gson();
        String json = gson.toJson(data);
        prefsEditor.putString("data", json);
        prefsEditor.commit();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        stopService(new Intent(getBaseContext(), GpsServices.class));
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        optionsMenu = menu;
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_refresh:
                onRefreshClick();
                return true;
            case R.id.action_settings:
                startActivity(new Intent(this, Settings.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onLocationChanged(Location location) {
        if (location.hasAccuracy()) {
            double acc = location.getAccuracy();
            String units;
            if (sharedPreferences.getBoolean("miles_per_hour", false)) {
                units = "ft";
                acc *= 3.28084;
            } else {
                units = "m";
            }

            SpannableString s = new SpannableString(String.format("%.0f %s", acc, units));
            s.setSpan(new RelativeSizeSpan(0.75f), s.length() - units.length() - 1, s.length(), 0);
            accuracy.setText(s);

            if (firstfix) {
                toolbar.setTitle(R.string.app_name);
                fab.show();
                if (!data.isRunning() && data.getMaxSpeed() != 0) {
                    optionsMenu.findItem(R.id.action_refresh).setVisible(true);
                }
                firstfix = false;
            }
        } else {
            firstfix = true;
        }

        if (location.hasSpeed()) {
            //progressBarCircularIndeterminate.setVisibility(View.GONE);
            double speed = location.getSpeed() * 3.6;
            String units;
            if (sharedPreferences.getBoolean("miles_per_hour", false)) { // Convert to MPH
                speed *= 0.62137119;
                units = "mi/h";
            } else {
                units = "km/h";
            }
            SpannableString s = new SpannableString(String.format("%.0f %s", speed, units));
            s.setSpan(new RelativeSizeSpan(0.25f), s.length() - units.length() - 1, s.length(), 0);
            currentSpeed.setText(s);
        }

    }

    /*@Override
    public void onGpsStatusChanged(int event) {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        switch (event) {
            case GpsStatus.GPS_EVENT_SATELLITE_STATUS:

                GpsStatus gpsStatus = mLocationManager.getGpsStatus(null);
                int satsInView = 0;
                int satsUsed = 0;
                Iterable<GpsSatellite> sats = gpsStatus.getSatellites();
                for (GpsSatellite sat : sats) {
                    satsInView++;
                    if (sat.usedInFix()) {
                        satsUsed++;
                    }
                }

                satellite.setText(String.format("%d/%d", satsUsed, satsInView));
                if (satsUsed == 0) {
                    fab.setImageDrawable(getResources().getDrawable(R.drawable.ic_action_play));
                    data.setRunning(false);
                    toolbar.setTitle(R.string.app_name);
                    stopService(new Intent(getBaseContext(), GpsServices.class));
                    fab.hide();
                    optionsMenu.findItem(R.id.action_refresh).setVisible(false);

                    accuracy.setText(String.format("%.0f %s", 0.0f,
                                    sharedPreferences.getBoolean("miles_per_hour", false) ? "ft" : "m"));

                    toolbar.setTitle(R.string.waiting_for_fix);
                    firstfix = true;
                }
                break;

            case GpsStatus.GPS_EVENT_STOPPED:
                if (!mLocationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
                    showGpsDisabledDialog();
                }
                break;
            case GpsStatus.GPS_EVENT_FIRST_FIX:
                break;
        }
    }*/
    
    @Override
    public void onStatusChanged(int event) {
        
        switch (event) {
            case 0: {
                if (!mPositioningStatus.isProviderEnabled()) {
                   fab.hide();
                   data.setRunning(false);
                   Toast.makeText(getApplicationContext(), getResources().getString(R.string.gps_disabled), Toast.LENGTH_SHORT)
                   .show();
                }
                break;
            }
            case 1:
                break;
            case 2: {
                satellite.setText(String.format("%d/%d", mPositioningStatus.getSatellitesUsed(),
                                        mPositioningStatus.getSatelliteCount()));
        
                if (mPositioningStatus.getSatellitesUsed() == 0) {
                    fab.setImageDrawable(getResources().getDrawable(R.drawable.ic_action_play));
                    fab.hide();
                    data.setRunning(false);
                    stopService(new Intent(getBaseContext(), GpsServices.class));
                    optionsMenu.findItem(R.id.action_refresh).setVisible(false);

                    accuracy.setText(String.format("%.0f %s", 0.0f,
                                    sharedPreferences.getBoolean("miles_per_hour", false) ? "ft" : "m"));

                    toolbar.setTitle(R.string.waiting_for_fix);
                    firstfix = true;
                }
                break;
            }
        }
    }

    public void showGpsDisabledDialog(){
        new MaterialAlertDialogBuilder(this)
                .setTitle(getResources().getString(R.string.gps_disabled))
                .setMessage(getResources().getString(R.string.please_enable_gps))
                .setPositiveButton(getResources().getString(R.string.accept), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        startActivity(new Intent("android.settings.LOCATION_SOURCE_SETTINGS"));
                    }
                })
                .show();
    }

    public void resetData(){
        fab.setImageDrawable(getResources().getDrawable(R.drawable.ic_action_play));
        optionsMenu.findItem(R.id.action_refresh).setVisible(false);

        maxSpeed.setText("");

        averageSpeed.setText("");

        distance.setText("---");

        time.stop();
        time.setBase(SystemClock.elapsedRealtime());

        data = new Data(onGpsServiceUpdate);
    }

    public static Data getData() {
        return data;
    }

    @Override
    public void onBackPressed(){
        Intent a = new Intent(Intent.ACTION_MAIN);
        a.addCategory(Intent.CATEGORY_HOME);
        a.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(a);
    }

    @Override
    public void onStatusChanged(String s, int i, Bundle bundle) {}

    @Override
    public void onProviderEnabled(String s) {}

    @Override
    public void onProviderDisabled(String s) {}
    
    private void createPositioningStatus() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            mPositioningStatus = new GnssStatusNg(this, this);
        }
        else {
            mPositioningStatus = new GpsStatusLegacy(this, this);
        }
    }
}

File: app/src/main/java/fly/speedmeter/grub/MainActivity.kt
package fly.speedmeter.grub

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.content.SharedPreferences
import android.Manifest
import android.os.Bundle
import android.os.IBinder
import android.os.Message
import android.os.Messenger
import android.os.RemoteException
import android.os.SystemClock
import android.text.format.DateUtils
import android.text.SpannableString
import android.text.style.RelativeSizeSpan
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import android.widget.Toast

import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.ActivityCompat
import androidx.core.content.edit
import androidx.core.content.ContextCompat
import androidx.core.content.PermissionChecker

import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.floatingactionbutton.FloatingActionButton


class MainActivity : AppCompatActivity() {

    private var mService: Messenger? = null

    private var bound: Boolean = false

    private var isFirstFix = false

    private var isRunning = false

    private var time: Long = 0L

    private val mMessenger: Messenger = Messenger(CustomHandler(this::handleMessage))

    private lateinit var tvSatelliteData: TextView

    private lateinit var tvAccuracyData: TextView

    private lateinit var tvCurrentSpeed: TextView

    private lateinit var tvMaxSpeedData: TextView

    private lateinit var tvAverageSpeedData: TextView

    private lateinit var tvAltitudeData: TextView

    private lateinit var tvDistance: TextView

    private lateinit var fabStart: FloatingActionButton

    private var menuOptions: Menu? = null

    private lateinit var mToolbar: MaterialToolbar

    private lateinit var tvTime: TextView

    private lateinit var mSharedPreferences: SharedPreferences

    private val mConnection = object : ServiceConnection {

        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            mService = Messenger(service)
            bound = true

            var msg = Message.obtain(null, REGISTER)
            msg.replyTo = mMessenger
            sendMessage(msg)

        }

        override fun onServiceDisconnected(className: ComponentName) {
            mService = null
            bound = false
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        mSharedPreferences = androidx.preference.PreferenceManager.getDefaultSharedPreferences(applicationContext)

        tvSatelliteData = findViewById(R.id.satelliteData)

        tvAccuracyData = findViewById(R.id.accuracyData)

        tvCurrentSpeed = findViewById(R.id.currentSpeed)

        tvMaxSpeedData = findViewById(R.id.maxSpeedData)

        tvAverageSpeedData = findViewById(R.id.averageSpeedData)

        tvAltitudeData = findViewById(R.id.altitudeData)

        tvDistance = findViewById(R.id.distance)
        tvDistance.text = "---"

        fabStart = findViewById(R.id.fab)
        fabStart.setVisibility(View.INVISIBLE)

        fabStart.setOnClickListener {
            if (!isRunning) {
                fabStart.setImageDrawable(ContextCompat.getDrawable(this, R.drawable.ic_action_pause))
                menuOptions?.findItem(R.id.action_refresh)?.setVisible(false)
                isRunning = true
            }
            else {
                fabStart.setImageDrawable(ContextCompat.getDrawable(this, R.drawable.ic_action_play))
                mToolbar.setTitle(R.string.app_name)
                menuOptions?.findItem(R.id.action_refresh)?.setVisible(true)
                isRunning = false
            }
            sendMessage(Message.obtain(null, RUNNING_UPDATE, isRunning))
        }

        mToolbar = findViewById(R.id.toolbar)
        setSupportActionBar(mToolbar)

        tvTime = findViewById(R.id.time)
        
        val theme = mSharedPreferences.getString("theme", "system")
        
        when (theme) {
            "light" -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            "dark" -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        }

        window.addFlags(android.view.WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        checkLocationPermission()
    }

    override fun onStart() {
        super.onStart()
        Intent(baseContext, GpsServices::class.java).also { intent ->
            startService(intent)
        }

        Intent(this, GpsServices::class.java).also { intent ->
            bindService(intent, mConnection, Context.BIND_AUTO_CREATE)
        }
    }

    override fun onStop() {
        super.onStop()

        unboundService()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        menuOptions = menu
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_refresh -> {
                sendMessage(Message.obtain(null, RESET))
                return true
            }
            R.id.action_exit -> {
                exitApplication()
                return true
            }
            R.id.action_settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
                return true
            }
            else -> {
                return super.onOptionsItemSelected(item)
            }
        }
    }

    override fun onBackPressed() {
        Intent(Intent.ACTION_MAIN).also { intent ->
            intent.addCategory(Intent.CATEGORY_HOME)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            startActivity(intent)
        }
    }

    fun showLocationDisabledDialog() {
        MaterialAlertDialogBuilder(this)
        .setMessage(resources.getString(R.string.please_enable_gps))
        .setPositiveButton(resources.getString(R.string.accept)) { _, _, ->
            startActivity(Intent("android.settings.LOCATION_SOURCE_SETTINGS"))
        }
        .show()
    }

    fun checkLocationPermission() {
        if (PermissionChecker.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
            != PermissionChecker.PERMISSION_GRANTED
            && PermissionChecker.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION)
            != PermissionChecker.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 1337)
        }
    }

    fun sendMessage(msg: Message) {
        try {
            mService?.send(msg)
        }
        catch (e: RemoteException) {
            e.printStackTrace()
        }
    }

    fun handleMessage(msg: Message) {
        when (msg.what) {
            DATA_UPDATE -> handleDataUpdate(msg.obj as PositioningData)
            GPS_DISABLED -> showLocationDisabledDialog()
        }
    }

    fun handleDataUpdate(data_: PositioningData) {
        isFirstFix = data_.isFirstTime
        handleRunningUpdate(data_.isRunning)
        handleTimeUpdate(data_.time)
        handleAccuracyUpdate(data_.accuracy)
        handleSatellitesUpdate(data_.satellitesUsed, data_.satellites)
        handleSpeedUpdate(data_.currentSpeed, data_.maxSpeed, calculateAverageSpeed(data_))
        handleAltitudeUpdate(data_.altitude)
        handleDistanceUpdate(data_.distance)
    }

    fun handleSatellitesUpdate(satellitesUsed: Int, satelliteCount: Int) {
        tvSatelliteData.text = "$satellitesUsed/$satelliteCount"

        if (satellitesUsed == 0) {
            fabStart.setImageDrawable(ContextCompat.getDrawable(this, R.drawable.ic_action_play))
            fabStart.hide()

            menuOptions?.findItem(R.id.action_refresh)?.setVisible(false)

            mToolbar.setTitle(R.string.waiting_for_fix)
        }
    }

    fun handleAccuracyUpdate(accuracy: Float) {
        var acc = accuracy
        if (acc != -1.0f) {
            val units = if (mSharedPreferences.getBoolean("imperial", false)) {
                acc *= 3.28084f
                "ft"
            } else {
                "m"
            }

            var span = SpannableString("%.0f %s".format(acc, units))
            span.setSpan(RelativeSizeSpan(0.75f), span.length - units.length - 1, span.length, 0)
            tvAccuracyData.text = span

            if (isFirstFix) {
                mToolbar.setTitle(R.string.app_name)
                fabStart.show()
                isFirstFix = false
            }
        }
        else {
            isFirstFix = true
        }
    }

    fun handleSpeedUpdate(speed: Float, maxSpeed: Float, averageSpeed: Float) {
        //TODO: hide speed views
        if (speed == -1.0f) {
            return
        }

        // convert from m/s to km/h
        var spd = speed * 3.6f
        var maxSpd = maxSpeed * 3.6f
        var avgSpd = averageSpeed

        val units = if (mSharedPreferences.getBoolean("imperial", false)) {
            spd *= 0.62137119f
            maxSpd *= 0.62137119f
            avgSpd *= 0.62137119f
            "mph"
        } else {
            "km/h"
        }

        var span = SpannableString("%.0f %s".format(spd, units))
        span.setSpan(RelativeSizeSpan(0.25f), span.length - units.length - 1, span.length, 0)
        tvCurrentSpeed.text = span

        span = SpannableString("%.0f %s".format(maxSpd, units))
        span.setSpan(RelativeSizeSpan(0.5f), span.length - units.length - 1, span.length, 0)
        tvMaxSpeedData.text = span

        span = SpannableString("%.0f %s".format(avgSpd, units))
        span.setSpan(RelativeSizeSpan(0.5f), span.length - units.length - 1, span.length, 0)
        tvAverageSpeedData.text = span
    }

    fun handleAltitudeUpdate(altitude: Double) {
        if (altitude == -1.0) {
            return
        }

        var altd = altitude

        val units = if (mSharedPreferences.getBoolean("imperial", false)) {
            altd *= 3.28084
            "ft"
        } else {
            "m"
        }

        var span = SpannableString("%.0f %s".format(altd, units))
        span.setSpan(RelativeSizeSpan(0.5f), span.length - units.length - 1, span.length, 0)
        tvAltitudeData.text = span
    }

    fun handleDistanceUpdate(distance: Double) {
        var dist = distance

        val units = if (mSharedPreferences.getBoolean("imperial", false)) {
            if (dist >= 1700) {
                dist *= 0.0006213712
                "mi"
            } else {
                dist *= 3.28084
                "ft"
            }
        } else {
            if (dist >= 1000.0) {
                dist /= 1000.0
                "km"
            }
            else {
                "m"
            }
        }

        var span = SpannableString("%.0f %s".format(dist, units))
        span.setSpan(RelativeSizeSpan(0.5f), span.length - units.length - 1, span.length, 0)
        tvDistance.text = span
    }

    fun handleRunningUpdate(running: Boolean) {
        isRunning = running

        if (isFirstFix) {
            return
        }

        if (isRunning) {
            fabStart.setImageDrawable(ContextCompat.getDrawable(this, R.drawable.ic_action_pause))
        }
        else {
            fabStart.setImageDrawable(ContextCompat.getDrawable(this, R.drawable.ic_action_play))
        }

        if (fabStart.visibility == View.INVISIBLE) {
            fabStart.show()
        }

    }

    fun handleTimeUpdate(time_: Long) {
        time = time_

        tvTime.text = DateUtils.formatElapsedTime(time)
    }

    fun calculateAverageSpeed(data_: PositioningData): Float {
        val autoAvg = mSharedPreferences.getBoolean("auto_average", false)

        val motionTime = if (autoAvg) {
            time - data_.timeStopped
        } else {
            time
        }

        val avgSpeed: Float = if (motionTime <= 0.0f) {
            0.0f
        } else {
            (data_.distance.toFloat() / motionTime.toFloat() * 3.6f)
        }

        return avgSpeed
    }

    fun unboundService() {
        if (bound) {
            if (mService != null) {
                sendMessage(Message.obtain(null, UNREGISTER))
            }

            unbindService(mConnection)
            bound = false
        }
    }

    fun exitApplication() {
        unboundService()

        Intent(applicationContext, GpsServices::class.java).also { intent ->
            stopService(intent)
        }

        finish()
    }

}
